src/dump.c

```diff
@@ -2346,14 +2349,21 @@ static void jl_init_restored_modules(jl_array_t *init_order)
 static void jl_prune_type_cache(jl_svec_t *cache)
 {
     size_t l = jl_svec_len(cache), ins = 0, i;
-    for(i=0; i < l; i++) {
+    for (i = 0; i < l; i++) {
         jl_value_t *ti = jl_svecref(cache, i);
-        if (ti == NULL) break;
+        if (ti == NULL)
+            break;
         if (ptrhash_get(&backref_table, ti) != HT_NOTFOUND || jl_get_llvm_gv(ti) != 0)
             jl_svecset(cache, ins++, ti);
+        else if (jl_is_datatype(ti)) {
+            jl_value_t *singleton = ((jl_datatype_t*)ti)->instance;
+            if (singleton && (ptrhash_get(&backref_table, singleton) != HT_NOTFOUND || jl_get_llvm_gv(singleton) != 0))
```

Comment:
probably only the `jl_get_llvm_gv` is needed. I just didn't feel like trying to verify / assert that